File: scripts/generate-user-guide-pdf.mjs
import fs from "node:fs";
import path from "node:path";
import PDFDocument from "pdfkit";

const rootDir = process.cwd();
const defaultSource = path.join(rootDir, "docs", "user-guide", "GST_BILLING_USER_GUIDE_SOURCE.md");
const defaultOutput = path.join(rootDir, "docs", "user-guide", "output", "GST_Billing_User_Guide_Draft.pdf");

const sourcePath = path.resolve(process.argv[2] || defaultSource);
const outputPath = path.resolve(process.argv[3] || defaultOutput);

if (!fs.existsSync(sourcePath)) {
  console.error(`Source file not found: ${sourcePath}`);
  process.exit(1);
}

fs.mkdirSync(path.dirname(outputPath), { recursive: true });

const markdown = fs.readFileSync(sourcePath, "utf8");
const sourceDir = path.dirname(sourcePath);

const outputStream = fs.createWriteStream(outputPath);

const doc = new PDFDocument({
  size: "A4",
  margin: 54,
  autoFirstPage: false,
  bufferPages: true,
  info: {
    Title: "GST Billing User Guide",
    Author: "Codex",
    Subject: "End-to-end user tutorial for GST Billing",
  },
});

doc.pipe(outputStream);

const colors = {
  ink: "#162231",
  muted: "#667991",
  accent: "#1d6d96",
  bronze: "#99652a",
  border: "#d9e1eb",
  panel: "#f5f7fb",
  darkPanel: "#1a2535",
  white: "#ffffff",
};

function stripMarkdown(text) {
  return text
    .replace(/\[(.*?)\]\((.*?)\)/g, "$1")
    .replace(/`([^`]+)`/g, "$1")
    .replace(/\*\*(.*?)\*\*/g, "$1")
    .replace(/\*(.*?)\*/g, "$1")
    .trim();
}

function ensureSpace(height = 24) {
  if (doc.y + height <= doc.page.height - doc.page.margins.bottom - 20) return;
  doc.addPage();
}

function writeHeading(level, text) {
  const clean = stripMarkdown(text);
  const size = level === 1 ? 27 : level === 2 ? 19 : level === 3 ? 14 : 12;
  ensureSpace(level === 1 ? 80 : 42);
  doc.moveDown(level === 1 ? 0.7 : 0.45);
  doc.font("Helvetica-Bold").fontSize(size).fillColor(colors.ink).text(clean, {
    width: doc.page.width - doc.page.margins.left - doc.page.margins.right,
  });
  if (level <= 2) {
    doc.moveDown(0.2);
    doc
      .strokeColor(colors.border)
      .lineWidth(1)
      .moveTo(doc.page.margins.left, doc.y + 4)
      .lineTo(doc.page.width - doc.page.margins.right, doc.y + 4)
      .stroke();
    doc.moveDown(0.5);
  } else {
    doc.moveDown(0.25);
  }
}

function writeParagraph(text) {
  const clean = stripMarkdown(text);
  if (!clean) {
    doc.moveDown(0.35);
    return;
  }
  ensureSpace(42);
  doc.font("Helvetica").fontSize(11).fillColor(colors.ink).text(clean, {
    width: doc.page.width - doc.page.margins.left - doc.page.margins.right,
    lineGap: 3,
  });
  doc.moveDown(0.45);
}

function writeBullet(text) {
  const clean = stripMarkdown(text);
  ensureSpace(28);
  const startX = doc.page.margins.left;
  const startY = doc.y + 5;
  doc.circle(startX + 4, startY, 2.3).fill(colors.accent);
  doc
    .fillColor(colors.ink)
    .font("Helvetica")
    .fontSize(11)
    .text(clean, startX + 14, doc.y, {
      width: doc.page.width - doc.page.margins.left - doc.page.margins.right - 14,
      lineGap: 3,
    });
  doc.moveDown(0.22);
}

function writeDivider() {
  ensureSpace(18);
  doc
    .strokeColor(colors.border)
    .lineWidth(1)
    .moveTo(doc.page.margins.left, doc.y + 2)
    .lineTo(doc.page.width - doc.page.margins.right, doc.y + 2)
    .stroke();
  doc.moveDown(0.65);
}

function writeImagePlaceholder(label, resolvedPath) {
  ensureSpace(220);
  const boxWidth = doc.page.width - doc.page.margins.left - doc.page.margins.right;
  const boxHeight = 180;
  const x = doc.page.margins.left;
  const y = doc.y;
  doc.roundedRect(x, y, boxWidth, boxHeight, 14).fillAndStroke(colors.panel, colors.border);
  doc
    .fillColor(colors.muted)
    .font("Helvetica-Bold")
    .fontSize(12)
    .text("Screenshot placeholder", x + 18, y + 22, { width: boxWidth - 36, align: "center" });
  doc
    .font("Helvetica")
    .fontSize(10)
    .text(stripMarkdown(label) || "Image", x + 26, y + 58, { width: boxWidth - 52, align: "center" });
  doc
    .fontSize(9)
    .fillColor(colors.muted)
    .text(`Expected asset: ${path.relative(rootDir, resolvedPath)}`, x + 26, y + 92, {
      width: boxWidth - 52,
      align: "center",
    });
  doc.y += boxHeight + 16;
}

function writeImage(label, imagePath) {
  const resolvedPath = path.resolve(sourceDir, imagePath);
  if (!fs.existsSync(resolvedPath)) {
    writeImagePlaceholder(label, resolvedPath);
    return;
  }
  ensureSpace(250);
  doc.image(resolvedPath, {
    fit: [doc.page.width - doc.page.margins.left - doc.page.margins.right, 240],
    align: "center",
    valign: "center",
  });
  doc.moveDown(0.4);
  doc.font("Helvetica-Oblique").fontSize(9).fillColor(colors.muted).text(stripMarkdown(label), {
    align: "center",
  });
  doc.moveDown(0.8);
}

function renderCover() {
  doc.addPage();
  const pageWidth = doc.page.width - doc.page.margins.left - doc.page.margins.right;
  const x = doc.page.margins.left;
  const y = 76;

  doc.roundedRect(x, y, pageWidth, 170, 24).fill(colors.darkPanel);
  doc
    .fillColor(colors.white)
    .font("Helvetica-Bold")
    .fontSize(13)
    .text("GST Billing", x + 28, y + 30);
  doc
    .font("Helvetica")
    .fontSize(10)
    .fillColor("#bfc9d9")
    .text("India-first billing operations", x + 28, y + 50);
  doc
    .fillColor(colors.white)
    .font("Helvetica-Bold")
    .fontSize(28)
    .text("End-to-End User Guide", x + 28, y + 84, { width: pageWidth - 56 });
  doc
    .font("Helvetica")
    .fontSize(11)
    .fillColor("#d7e2f0")
    .text("Draft PDF generated from the repository manuscript source.", x + 28, y + 126, {
      width: pageWidth - 56,
    });

  doc.moveDown(12);
  doc
    .fillColor(colors.ink)
    .font("Helvetica-Bold")
    .fontSize(18)
    .text("What this guide covers");
  doc.moveDown(0.6);

  [
    "Public entry, onboarding, and login",
    "Sales, purchases, inventory, collections, GST, accounting, POS, and field sales",
    "Reports, settings, governance, and internal admin workflows",
    "Step-by-step operator guidance with image placeholders ready for final screenshots",
  ].forEach((item) => writeBullet(item));

  doc.moveDown(0.8);
  doc
    .fillColor(colors.muted)
    .font("Helvetica")
    .fontSize(10)
    .text(`Generated on ${new Date().toISOString().slice(0, 10)}`, { align: "left" });
}

function renderBody() {
  doc.addPage();
  const lines = markdown.split(/\r?\n/);
  for (const line of lines) {
    const trimmed = line.trim();

    if (!trimmed) {
      doc.moveDown(0.35);
      continue;
    }

    const imageMatch = trimmed.match(/^!\[(.*?)\]\((.*?)\)$/);
    if (imageMatch) {
      writeImage(imageMatch[1], imageMatch[2]);
      continue;
    }

    if (trimmed === "---") {
      writeDivider();
      continue;
    }

    const headingMatch = trimmed.match(/^(#{1,6})\s+(.*)$/);
    if (headingMatch) {
      writeHeading(headingMatch[1].length, headingMatch[2]);
      continue;
    }

    const orderedMatch = trimmed.match(/^\d+\.\s+(.*)$/);
    if (orderedMatch) {
      writeBullet(orderedMatch[1]);
      continue;
    }

    const bulletMatch = trimmed.match(/^-\s+(.*)$/);
    if (bulletMatch) {
      writeBullet(bulletMatch[1]);
      continue;
    }

    writeParagraph(trimmed);
  }
}

renderCover();
renderBody();

const range = doc.bufferedPageRange();
for (let i = 0; i < range.count; i += 1) {
  doc.switchToPage(i);
  doc
    .font("Helvetica")
    .fontSize(9)
    .fillColor(colors.muted)
    .text(`Page ${i + 1}`, 0, doc.page.height - 36, {
      width: doc.page.width,
      align: "center",
    });
}

doc.end();

await new Promise((resolve, reject) => {
  outputStream.on("finish", resolve);
  outputStream.on("error", reject);
});

console.log(`User guide PDF generated: ${outputPath}`);

File: scripts/generate_user_guide_pdf.py
#!/usr/bin/env python3

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import List

from reportlab.lib import colors
from reportlab.lib.enums import TA_CENTER, TA_JUSTIFY
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, StyleSheet1, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.platypus import (
    BaseDocTemplate,
    Frame,
    Image,
    KeepTogether,
    ListFlowable,
    ListItem,
    PageBreak,
    PageTemplate,
    Paragraph,
    SimpleDocTemplate,
    Spacer,
    Table,
    TableStyle,
)


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "docs" / "user-guide" / "GST_BILLING_USER_GUIDE_SOURCE.md"
OUTPUT = ROOT / "output" / "pdf" / "GST_Billing_User_Guide_Manual.pdf"
TMP_DIR = ROOT / "tmp" / "pdfs"


def build_styles() -> StyleSheet1:
    styles = getSampleStyleSheet()
    styles.add(
        ParagraphStyle(
            name="BodyCopy",
            parent=styles["BodyText"],
            fontName="Helvetica",
            fontSize=10.5,
            leading=15,
            textColor=colors.HexColor("#203044"),
            spaceAfter=6,
            alignment=TA_JUSTIFY,
        )
    )
    styles.add(
        ParagraphStyle(
            name="CoverKicker",
            parent=styles["BodyText"],
            fontName="Helvetica-Bold",
            fontSize=11,
            leading=14,
            textColor=colors.HexColor("#D9E5F3"),
            spaceAfter=8,
            alignment=TA_CENTER,
        )
    )
    styles.add(
        ParagraphStyle(
            name="CoverTitle",
            parent=styles["Title"],
            fontName="Helvetica-Bold",
            fontSize=26,
            leading=30,
            alignment=TA_CENTER,
            textColor=colors.white,
            spaceAfter=12,
        )
    )
    styles.add(
        ParagraphStyle(
            name="CoverText",
            parent=styles["BodyText"],
            fontName="Helvetica",
            fontSize=11,
            leading=16,
            alignment=TA_CENTER,
            textColor=colors.HexColor("#E6EEF8"),
        )
    )
    styles.add(
        ParagraphStyle(
            name="CoverNote",
            parent=styles["BodyText"],
            fontName="Helvetica",
            fontSize=10.5,
            leading=14,
            alignment=TA_CENTER,
            textColor=colors.HexColor("#E6EEF8"),
        )
    )
    styles.add(
        ParagraphStyle(
            name="SectionTitle",
            parent=styles["Heading1"],
            fontName="Helvetica-Bold",
            fontSize=22,
            leading=28,
            textColor=colors.HexColor("#182434"),
            spaceBefore=8,
            spaceAfter=10,
        )
    )
    styles.add(
        ParagraphStyle(
            name="ChapterTitle",
            parent=styles["Heading2"],
            fontName="Helvetica-Bold",
            fontSize=16,
            leading=22,
            textColor=colors.HexColor("#182434"),
            spaceBefore=10,
            spaceAfter=8,
        )
    )
    styles.add(
        ParagraphStyle(
            name="MinorTitle",
            parent=styles["Heading3"],
            fontName="Helvetica-Bold",
            fontSize=12.5,
            leading=17,
            textColor=colors.HexColor("#23415A"),
            spaceBefore=6,
            spaceAfter=6,
        )
    )
    styles.add(
        ParagraphStyle(
            name="Caption",
            parent=styles["BodyText"],
            fontName="Helvetica-Oblique",
            fontSize=8.5,
            leading=11,
            textColor=colors.HexColor("#6D7E92"),
            alignment=TA_CENTER,
            spaceBefore=3,
            spaceAfter=12,
        )
    )
    styles.add(
        ParagraphStyle(
            name="Placeholder",
            parent=styles["BodyText"],
            fontName="Helvetica-Bold",
            fontSize=11,
            leading=14,
            alignment=TA_CENTER,
            textColor=colors.HexColor("#5E6F84"),
        )
    )
    return styles


STYLES = build_styles()


def clean_inline(text: str) -> str:
    text = re.sub(r"`([^`]+)`", r"<font face='Helvetica-Bold'>\1</font>", text)
    text = re.sub(r"\*\*(.+?)\*\*", r"<b>\1</b>", text)
    text = re.sub(r"\*(.+?)\*", r"<i>\1</i>", text)
    text = re.sub(r"\[(.*?)\]\((.*?)\)", r"<u>\1</u>", text)
    replacements = {
        "<u>": "%%U_OPEN%%",
        "</u>": "%%U_CLOSE%%",
        "<b>": "%%B_OPEN%%",
        "</b>": "%%B_CLOSE%%",
        "<i>": "%%I_OPEN%%",
        "</i>": "%%I_CLOSE%%",
        "<font face='Helvetica-Bold'>": "%%F_OPEN%%",
        "</font>": "%%F_CLOSE%%",
    }
    for original, token in replacements.items():
        text = text.replace(original, token)
    text = text.replace("&", "&amp;")
    text = text.replace("<", "&lt;").replace(">", "&gt;")
    for original, token in replacements.items():
        text = text.replace(token, original)
    return text


def image_flowable(path: Path, caption: str):
    if path.exists():
      img = Image(str(path))
      img._restrictSize(165 * mm, 105 * mm)
      return KeepTogether([img, Spacer(1, 3), Paragraph(clean_inline(caption), STYLES["Caption"])])

    placeholder = Table(
        [[Paragraph("Screenshot placeholder", STYLES["Placeholder"])], [Paragraph(clean_inline(caption), STYLES["Caption"])]],
        colWidths=[165 * mm],
        rowHeights=[22 * mm, 18 * mm],
    )
    placeholder.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, -1), colors.HexColor("#F2F6FB")),
                ("BOX", (0, 0), (-1, -1), 1, colors.HexColor("#D5DFEA")),
                ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
                ("ALIGN", (0, 0), (-1, -1), "CENTER"),
                ("LEFTPADDING", (0, 0), (-1, -1), 12),
                ("RIGHTPADDING", (0, 0), (-1, -1), 12),
                ("TOPPADDING", (0, 0), (-1, -1), 10),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 10),
            ]
        )
    )
    return KeepTogether([placeholder, Spacer(1, 10)])


def make_cover_story() -> List:
    story: List = []
    story.append(Spacer(1, 12 * mm))
    hero_panel = Table(
        [
            [Paragraph("GST Billing", STYLES["CoverKicker"])],
            [Paragraph("End-to-End User Guide", STYLES["CoverTitle"])],
            [
                Paragraph(
                    "A practical operator manual for billing, purchases, inventory, GST, accounting, collections, POS, field sales, settings, and internal admin workflows.",
                    STYLES["CoverText"],
                )
            ],
        ],
        colWidths=[170 * mm],
    )
    hero_panel.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, -1), colors.HexColor("#182434")),
                ("BOX", (0, 0), (-1, -1), 0.5, colors.HexColor("#182434")),
                ("LEFTPADDING", (0, 0), (-1, -1), 18),
                ("RIGHTPADDING", (0, 0), (-1, -1), 18),
                ("TOPPADDING", (0, 0), (-1, -1), 18),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 18),
                ("ALIGN", (0, 0), (-1, -1), "CENTER"),
                ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
            ]
        )
    )
    story.append(hero_panel)
    story.append(Spacer(1, 8 * mm))
    cover_image = ROOT / "docs" / "user-guide" / "assets" / "public-landing.png"
    story.append(image_flowable(cover_image, "Public landing page and product positioning"))
    story.append(PageBreak())
    return story


def parse_markdown(source_path: Path) -> List:
    story: List = []
    lines = source_path.read_text(encoding="utf-8").splitlines()
    pending_bullets: List[str] = []

    def flush_bullets():
        nonlocal pending_bullets
        if not pending_bullets:
            return
        list_items = [
            ListItem(Paragraph(clean_inline(item), STYLES["BodyCopy"]), leftIndent=6) for item in pending_bullets
        ]
        story.append(
            ListFlowable(
                list_items,
                bulletType="bullet",
                bulletFontName="Helvetica-Bold",
                bulletColor=colors.HexColor("#1D6D96"),
                leftIndent=16,
            )
        )
        story.append(Spacer(1, 4))
        pending_bullets = []

    for raw in lines:
        line = raw.rstrip()
        stripped = line.strip()

        if not stripped:
            flush_bullets()
            story.append(Spacer(1, 4))
            continue

        if stripped == "---":
            flush_bullets()
            story.append(Spacer(1, 3))
            divider = Table([[""]], colWidths=[170 * mm], rowHeights=[1])
            divider.setStyle(TableStyle([("BACKGROUND", (0, 0), (-1, -1), colors.HexColor("#D9E1EB"))]))
            story.append(divider)
            story.append(Spacer(1, 8))
            continue

        image_match = re.match(r"^!\[(.*?)\]\((.*?)\)$", stripped)
        if image_match:
            flush_bullets()
            image_path = (source_path.parent / image_match.group(2)).resolve()
            story.append(image_flowable(image_path, image_match.group(1)))
            continue

        heading = re.match(r"^(#{1,3})\s+(.*)$", stripped)
        if heading:
            flush_bullets()
            level = len(heading.group(1))
            text = clean_inline(heading.group(2))
            style = STYLES["SectionTitle"] if level == 1 else STYLES["ChapterTitle"] if level == 2 else STYLES["MinorTitle"]
            story.append(Paragraph(text, style))
            continue

        bullet = re.match(r"^(?:-|\d+\.)\s+(.*)$", stripped)
        if bullet:
            pending_bullets.append(bullet.group(1))
            continue

        flush_bullets()
        story.append(Paragraph(clean_inline(stripped), STYLES["BodyCopy"]))

    flush_bullets()
    return story


def on_page(canvas, doc):
    page_no = canvas.getPageNumber()
    width, height = A4
    canvas.saveState()
    canvas.setStrokeColor(colors.HexColor("#D8E1EB"))
    canvas.line(18 * mm, height - 14 * mm, width - 18 * mm, height - 14 * mm)
    canvas.setFont("Helvetica-Bold", 8)
    canvas.setFillColor(colors.HexColor("#5F7289"))
    canvas.drawString(18 * mm, height - 11.5 * mm, "GST Billing User Guide")
    canvas.setFont("Helvetica", 8)
    canvas.drawRightString(width - 18 * mm, height - 11.5 * mm, f"Page {page_no}")
    canvas.setStrokeColor(colors.HexColor("#D8E1EB"))
    canvas.line(18 * mm, 14 * mm, width - 18 * mm, 14 * mm)
    canvas.setFont("Helvetica", 7.5)
    canvas.setFillColor(colors.HexColor("#7A8A9E"))
    canvas.drawString(18 * mm, 9.5 * mm, "Generated from repository guide source")
    canvas.restoreState()


def build_pdf():
    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    TMP_DIR.mkdir(parents=True, exist_ok=True)

    doc = SimpleDocTemplate(
        str(OUTPUT),
        pagesize=A4,
        leftMargin=20 * mm,
        rightMargin=20 * mm,
        topMargin=20 * mm,
        bottomMargin=20 * mm,
        title="GST Billing User Guide",
        author="Codex",
    )

    story = []
    story.extend(make_cover_story())
    story.extend(parse_markdown(SOURCE))
    doc.build(story, onFirstPage=on_page, onLaterPages=on_page)


if __name__ == "__main__":
    build_pdf()
    print(f"Generated: {OUTPUT}")
